### PR TITLE
fix compact task data race

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>

### CODE REVIEW

File: internal/datanode/compaction_executor_test.go [hunk 1]
```diff
@@ -133,17 +133,13 @@ type mockCompactor struct {
 	isvalid       bool
 	alwaysWorking bool
 
-	wg sync.WaitGroup
+	done chan bool
```
Comment (congqixia) [inline internal/datanode/compaction_executor_test.go:136]: prefer `done chan struct{}` here

File: internal/datanode/compactor.go [hunk 2]
```diff
@@ -81,7 +80,7 @@ type compactionTask struct {
 	ctx    context.Context
 	cancel context.CancelFunc
 
-	wg           sync.WaitGroup
+	done         chan bool
```
Comment (congqixia) [inline internal/datanode/compactor.go:83]: ditto

File: internal/datanode/compaction_executor_test.go [hunk 3]
```diff
@@ -124,6 +123,7 @@ func newMockCompactor(isvalid bool) *mockCompactor {
 		ctx:     ctx,
 		cancel:  cancel,
 		isvalid: isvalid,
+		done:    make(chan struct{}, 1),
```
Comment (yah01) [inline internal/datanode/compaction_executor_test.go:126]: use a unbuffered channel will be better as we don't need to type "1" many times

File: internal/datanode/compaction_executor_test.go [hunk 4]
```diff
@@ -133,22 +133,13 @@ type mockCompactor struct {
...
-	mc.mu.Lock()
-	defer mc.mu.Unlock()
-	mc.wg.Add(1)
-}
-
 func (mc *mockCompactor) complete() {
-	mc.mu.Lock()
-	defer mc.mu.Unlock()
-	mc.wg.Done()
+	mc.done <- struct{}{}
```
Comment (yah01) [inline internal/datanode/compaction_executor_test.go:142]: just closing this channel, it works the same

File: internal/datanode/compaction_executor_test.go (hunk 3 above)
Comment (weiliu1031) [reply internal/datanode/compaction_executor_test.go:126]: with unbuffered channel, if call `done<- struct{}{}` after `<-done`, will cause deadlock

Comment (yah01) [reply internal/datanode/compaction_executor_test.go:126]: > with unbuffered channel, if call `done<- struct{}{}` after `<-done`, will cause deadlock

We can just close this channel, not to write anything into it

**Review by congqixia — APPROVED** [review]
@weiliu1031 please fix @yah01 's comment in next PRs
/lgtm

**Pull request merged**